### PR TITLE
Add CLI to perform blending with mismatching vicinities and the option to rename the cube

### PR DESCRIPTION
**Pull request by MoseleyS**
Part of https://github.com/MetOffice/improver_suite/issues/1610

To achieve the meta-data changes requested here (https://github.com/MetOffice/improver_suite/pull/1611#discussion_r1165375538), I have created a new CLI to call `weighted-blending`, then modify the meta-data of the resulting cube before it is returned.

The metadata changes are:
- File name: `f"precipitation_accumulation_in_variable_vicinity-PT{period:02d}H"`
- Cube name: `probability_of_lwe_thickness_of_precipitation_amount_in_variable_vicinity_above_threshold`
- Cube coordinate: `radius_of_vicinity`, Value: 10000 m, With the addition of this comment attribute in the nowcast period only: `comment: "Maximum"`

Implementing this required these changes to allow me to reuse existing code:
- Refactors `update_diagnostic_name` and `_update_cell_methods` from `cube_combiner.CubeMultiplier` to `metadata.amend.py`
- Refactors `add_vicinity_coordinate` from `spatial.OccurrenceWithinVicinity` plugin to `spatial.py`
- New unit tests for these refactored methods.
- Modification to allow the `interpret-metadata` CLI to accept the new metadata.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: improver/cli/blend_with_vicinity_and_rename.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    cycletime: str = None,
+    y0val: float = None,
+    ynval: float = None,
+    cval: float = None,
+    model_id_attr: str = None,
+    record_run_attr: str = None,
+    spatial_weights_from_mask=False,
+    fuzzy_length=20000.0,
+):
+    """Runs weighted blending, with additional options to handle mismatched in_vicinity coords
```
Comment (brhooper) [inline improver/cli/blend_with_vicinity_and_rename.py:58]: This line might be too long.

File: improver/cli/blend_with_vicinity_and_rename.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+            If calc_method is nonlinear and either y0val and ynval is not None.
+        RuntimeError:
+            If calc_method is dict and weights_dict is None.
+    """
+    from improver.cli import weighted_blending
+    from improver.utilities.spatial import update_name_and_vicinity_coord
+
+    if not new_name and not vicinity_radius:
+        raise ValueError(
+            "Neither new_name nor vicinity_radius have been specified. Use weighted-blending "
```
Comment (brhooper) [inline improver/cli/blend_with_vicinity_and_rename.py:151]: This line might be too long.

File: improver/metadata/amend.py [hunk 3]
```diff
@@ -131,3 +137,87 @@ def update_model_id_attr_attribute(
...
     return {model_id_attr: " ".join(sorted(set(attr_list)))}
+
+
+def update_diagnostic_name(source_cube: Cube, new_diagnostic_name: str, result: Cube):
+    """
+    Used for renaming the threshold coordinate and modifying cell methods
+    where necessary; excludes the in_vicinity component.
+
+    Args:
+        source_cube: An original cube before any processing took place. Can be the same cube as
```
Comment (brhooper) [inline improver/metadata/amend.py:148]: This line might be too long.

File: improver/metadata/probabilistic.py [hunk 4]
```diff
@@ -55,7 +55,7 @@ def probability_cube_name_regex(cube_name: str) -> Optional[Match]:
     regex = re.compile(
         "(probability_of_)"  # always starts this way
         "(?P<diag>.*?)"  # named group for the diagnostic name
-        "(_in_vicinity|)"  # optional group, may be empty
+        "(?P<vicinity>_in_vicinity|_in_variable_vicinity)?"  # optional group, may be empty
```
Comment (brhooper) [inline improver/metadata/probabilistic.py:58]: This line might be too long.

File: improver/utilities/spatial.py [hunk 5]
```diff
@@ -716,3 +688,76 @@ def transform_grid_to_lat_lon(cube: Cube) -> Tuple[ndarray, ndarray]:
     lats = points[..., 1]
 
     return lats, lons
+
+
+def update_name_and_vicinity_coord(cube: Cube, new_name: str, vicinity_radius: float):
+    """
+    Updates a cube with a new probabilistic-style name and replaces or adds a radius_of_vicinity
```
Comment (brhooper) [inline improver/utilities/spatial.py:695]: This line might be too long.

File: improver/utilities/spatial.py [hunk 6]
```diff
@@ -716,3 +688,76 @@ def transform_grid_to_lat_lon(cube: Cube) -> Tuple[ndarray, ndarray]:
...
+    coord with the specified radius.
+
+    Args:
+        cube: Cube to be updated in-place
+        new_name: The new name to be applied to the Cube, the threshold coord and any related
```
Comment (brhooper) [inline improver/utilities/spatial.py:700]: This line might be too long.

File: improver/utilities/spatial.py [hunk 7]
```diff
@@ -716,3 +688,76 @@ def transform_grid_to_lat_lon(cube: Cube) -> Tuple[ndarray, ndarray]:
...
+            cell methods.
+        vicinity_radius: The point value for the radius_of_vicinity coord. The units are assumed
```
Comment (brhooper) [inline improver/utilities/spatial.py:702]: This line might be too long.

File: improver/utilities/spatial.py [hunk 8]
```diff
@@ -716,3 +688,76 @@ def transform_grid_to_lat_lon(cube: Cube) -> Tuple[ndarray, ndarray]:
...
+            to be the same as the x and y spatial coords of the Cube
+
+    """
+    if new_name:
+        update_diagnostic_name(cube, new_name, cube)
+    if vicinity_radius:
+        # The cube blending will drop the radius_of_vicinity coord if the source cubes have
```
Comment (brhooper) [inline improver/utilities/spatial.py:709]: This line might be too long.

File: improver/utilities/spatial.py [hunk 9]
```diff
@@ -716,3 +688,76 @@ def transform_grid_to_lat_lon(cube: Cube) -> Tuple[ndarray, ndarray]:
...
+    Args:
+        cube:
+            Vicinity processed cube.
+        radius:
+            The radius as a physical distance (m) or number of grid points, the
+            value of which is recorded in the coordinate.
+        native_grid_point_radius:
+            True if radius is "number of grid points", else False
+        radius_is_max:
+            True if the specified radius represents a maximum value from the source data. A
```
Comment (brhooper) [inline improver/utilities/spatial.py:740]: This line might be too long.

File: improver_tests/metadata/test_amend.py [hunk 10]
```diff
@@ -243,5 +247,52 @@ def test_attribute_mismatch(self):
             update_model_id_attr_attribute([self.cube1, self.cube2], self.model_id_attr)
 
 
+@pytest.mark.parametrize("cell_method", (True, False))
+@pytest.mark.parametrize("probability_data", (True, False))
+@pytest.mark.parametrize("in_vicinity", (True, False))
+def test_update_diagnostic_name(cell_method, in_vicinity, probability_data):
+    """Make sure that the update_diagnostic_name method makes the expected changes in the expected
```
Comment (brhooper) [inline improver_tests/metadata/test_amend.py:254]: This docstring might have lines which are too long.

**Review by brhooper — CHANGES_REQUESTED** [review]
Thanks @MoseleyS, I think this is mostly fine.

The only comments I've made are related to code style. I think I recall being told that there was a restriction of having up to 88 characters per line, though I can't seem to find a source for this now. If this is not the case then feel free to ignore my comments and I'll approve the PR.

**Comment by MoseleyS** [comment]
> Thanks @MoseleyS, I think this is mostly fine.
> 
> The only comments I've made are related to code style. I think I recall being told that there was a restriction of having up to 88 characters per line, though I can't seem to find a source for this now. If this is not the case then feel free to ignore my comments and I'll approve the PR.

Flake8 currently enforces a line-limit of 100 characters, so I tend to go with that. We've had different limits in the past. I know that there are other lines in our code base that are between 88 and 100 characters, so I don't think we need worry about this.

**Review by brhooper — APPROVED** [review]

**Review by Katie-Howard — CHANGES_REQUESTED** [review]
I have looked through all the changes and they all look good to me 👍 However when I run the unit/acceptance tests I get one failure in improver_tests/acceptance/test_checksums.py::test_kgo_checksums. Can you just check that it doesn't do that for you? It could easily be something on my side, but it would be good to check. Thanks.

**Comment by MoseleyS** [comment]
@Katie-Howard: The tests all pass for me this morning.

**Review by Katie-Howard — APPROVED** [review]
I have had a look at the new acceptance test data and I think that it looks fine. So I'm happy with all the changes 👍

**Pull request merged**